back/app/models/navbar_item.rb

```diff
@@ -0,0 +1,106 @@
+class NavbarItem < ActiveRecord::Base
+  self.primary_key = 'id'
+  self.inheritance_column = :_type_disabled
+
+  TYPES = %i[home projects proposals events all_input custom].freeze
+
+  FEATURES = {
+    destroy: %w[custom].freeze,
+    reorder: %w[proposals events all_input custom].freeze,
+    show_or_hide: %w[proposals events all_input custom].freeze,
+    edit_navbar_title: %w[projects proposals events all_input custom].freeze,
+    edit_page_slug: %w[custom].freeze,
+    edit_page_title: %w[projects all_input custom].freeze,
+    edit_page_body: %w[custom].freeze
+  }.freeze
+
+  MAX_VISIBLE_ITEMS = 7
+  LAST_RESERVED_ORDERING = 1
+
+  belongs_to :page
+
+  validates_numericality_of(:ordering, greater_than_or_equal_to: 0)
+  validates_numericality_of(
+    :ordering, if: :visible,
+    greater_than: LAST_RESERVED_ORDERING,
+    less_than: MAX_VISIBLE_ITEMS
+  )
+  validate :title_connot_be_more_than_20_characters
+  validate :ordering_cannot_be_aside_from_others, on: :update
+  validate :cannot_reorder_reserved_items
+  validate :cannot_show_or_hide_reserved_items
+  validate :cannot_edit_titles
+  validate :list_of_visible_items_is_already_full
+
+  before_destroy :validate_before_destroy
+
+  def self.maximum_ordering
+    maximum(:ordering) || -1
+  end
+
+  private
+
+  def title_connot_be_more_than_20_characters
+    title_multiloc.each do |lang, title|
+      if title.size > 20
+        errors.add :title_multiloc, "Cannot be more than 20 characters (lang: #{lang}, size: #{title.size})"
+      end
+    end
+  end
+
+  def ordering_cannot_be_aside_from_others
+    if ordering > (self.class.where(visible: visible).maximum_ordering + 1)
+      errors.add(:ordering, "cannot be aside from others")
+    end
+  end
```

Comment:
From @kogre
> Have bad experience with such validations (that are not just about the current record, but depend on other records). These will work as desired in most scenarios, but will break when e.g. you create some items in bulk, delete an item in-between, etc.

In the rest of the code, we're using acts_as_list gem for it. It works pretty well, and by itself deals with list-related validations. Not sure whether you saw that, but if you did, any reason not to adopt it here?

Reply:
@kogre

Reply:
By the way, I don't really like rails validations. I think they are messy and don't allow to write clean code. I made use of them for navbar only because the project uses them everywhere.

I used to use `dry-validations`. They allow to separate validations from models and make everything much cleaner.

If you don't mind, I would like to use `dry-validations` for the next epic.

Reply:
You can maybe give a small demo in the next BE CDM next Thursday on dry-validations, to see if we find it worthwhile to try? Feel free to add it to the agenda.